libs/backend/src/cast_vote_records/export.ts

```diff
@@ -423,6 +432,113 @@ async function exportSignatureFileToUsbDrive(
+  directoryPath: string
+): Promise<void> {
+  const tempDirectoryPath = `${directoryPath}-temp`;
+  await fs.mkdir(tempDirectoryPath);
+  const fileNames = (
+    await fs.readdir(directoryPath, { withFileTypes: true })
+  ).map((entry) => {
+    assert(
+      entry.isFile(),
+      `Unexpected sub-directory ${entry.name} in ${directoryPath}`
+    );
+    return entry.name;
+  });
+  for (const fileName of fileNames) {
+    await fs.copyFile(
+      path.join(directoryPath, fileName),
+      path.join(tempDirectoryPath, fileName)
+    );
+  }
+  await fs.rm(directoryPath, { recursive: true });
+  await fs.rename(tempDirectoryPath, directoryPath);
+}
+
+/**
+ * File creation timestamps could reveal the order in which ballots were cast. To maintain
+ * voter privacy, every time a ballot is cast, we randomly select 1 or 2 previously added cast vote
+ * records and update their creation timestamps to the present. This approach is equivalent to
+ * moving random ballots to the top of a stack every time a new ballot is added to the stack, a
+ * kind of shuffling as we go.
+ *
+ * This shuffling as we go is irrelevant for full exports, where we already iterate over cast vote
+ * records in an order independent of creation timestamp.
+ */
+async function maintainVoterPrivacy(
+  exportContext: ExportContext,
+  mostRecentlyAddedCastVoteRecordId: string
+): Promise<void> {
+  const { usbMountPoint } = exportContext;
+
+  const exportDirectoryPath = path.join(
+    usbMountPoint,
+    getExportDirectoryPathRelativeToUsbMountPoint(exportContext)
+  );
+  const previouslyAddedCastVoteRecordIds = (
+    await fs.readdir(exportDirectoryPath, { withFileTypes: true })
+  )
+    .filter(
+      (entry) =>
+        entry.isDirectory() && entry.name !== mostRecentlyAddedCastVoteRecordId
+    )
+    .map((entry) => entry.name);
+
+  if (previouslyAddedCastVoteRecordIds.length === 0) {
+    return;
+  }
+
+  const castVoteRecordsToUpdate: string[] = [];
+  if (previouslyAddedCastVoteRecordIds.length === 1) {
+    if (getRandomInteger({ min: 1, max: 2 }) === 1) {
+      castVoteRecordsToUpdate.push(
+        assertDefined(previouslyAddedCastVoteRecordIds[0])
+      );
+    }
+  } else {
+    const oneOrTwo = getRandomInteger({ min: 1, max: 2 });
+    for (let i = 0; i < oneOrTwo; i += 1) {
+      const randomIndex = getRandomInteger({
+        min: 0,
+        max: previouslyAddedCastVoteRecordIds.length - 1,
+      });
+      castVoteRecordsToUpdate.push(
+        assertDefined(previouslyAddedCastVoteRecordIds[randomIndex])
+      );
+    }
+    // 25% of the time, move the most recently added cast vote record back to the top of the stack.
+    // This is meant to protect the privacy of the last voter, whose cast vote record would
+    // otherwise always have the earliest timestamp among the cluster of cast vote records with the
+    // latest timestamps.
+    if (getRandomInteger({ min: 1, max: 4 }) === 1) {
```

Comment:
I picked a 25% probability for this piece somewhat arbitrarily. Just trying to strike a balance between not having to copy too much but adding enough randomness that you can't know for sure which cast vote record belongs to the last voter.

There's an alternative way to go about this: probabilistically deciding whether we call `maintainVoterPrivacy` before or after creating the new cast vote record. But that makes `maintainVoterPrivacy` less self-contained and the overall code a bit harder to reason about in my opinion.

Reply:
Had a change of heart here 🙈 and decided that the alternative was in fact still easy enough to understand, while also being more performant since we never have to copy the just created cast vote record directory

https://github.com/votingworks/vxsuite/pull/3972/commits/7fc2c4f58daa618df81e24af3b08731c47aa226f